resilience4j-ratpack/src/main/java/io/github/resilience4j/ratpack/timelimiter/TimeLimiterMethodInterceptor.java

```diff
@@ -0,0 +1,154 @@
+    @Nullable
+    private TimeLimiterRegistry registry;
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    @Override
+    public Object invoke(MethodInvocation invocation) throws Throwable {
+        TimeLimiter annotation = invocation.getMethod().getAnnotation(TimeLimiter.class);
+        if (annotation == null) {
+            annotation = invocation.getMethod().getDeclaringClass()
+                .getAnnotation(TimeLimiter.class);
+        }
+        final RecoveryFunction<?> fallbackMethod = Optional
+            .ofNullable(createRecoveryFunction(invocation, annotation.fallbackMethod()))
+            .orElse(new DefaultRecoveryFunction<>());
+        if (registry == null) {
+            registry = TimeLimiterRegistry.ofDefaults();
+        }
+        io.github.resilience4j.timelimiter.TimeLimiter timeLimiter = registry
+            .timeLimiter(annotation.name());
+        Class<?> returnType = invocation.getMethod().getReturnType();
+        if (Promise.class.isAssignableFrom(returnType)) {
+            Promise<?> result = (Promise<?>) proceed(invocation);
+            if (result != null) {
+                TimeLimiterTransformer transformer = TimeLimiterTransformer.of(timeLimiter)
+                    .recover(fallbackMethod);
+                result = result.transform(transformer);
+            }
+            return result;
+        } else if (Flux.class.isAssignableFrom(returnType)) {
+            Flux<?> result = (Flux<?>) proceed(invocation);
+            if (result != null) {
+                TimeLimiterOperator operator = TimeLimiterOperator.of(timeLimiter);
+                result = fallbackMethod.onErrorResume(result.transform(operator));
+            }
+            return result;
+        } else if (Mono.class.isAssignableFrom(returnType)) {
+            Mono<?> result = (Mono<?>) proceed(invocation);
+            if (result != null) {
+                TimeLimiterOperator operator = TimeLimiterOperator.of(timeLimiter);
+                result = fallbackMethod.onErrorResume(result.transform(operator));
+            }
+            return result;
+        } else if (CompletionStage.class.isAssignableFrom(returnType)) {
+            ScheduledExecutorService scheduler = Execution.current().getController().getExecutor();
+            CompletableFuture<?> future = timeLimiter.executeCompletionStage(scheduler, () -> {
+                try {
+                    return (CompletionStage)proceed(invocation);
+                } catch (Throwable t) {
+                    final CompletableFuture<?> promise = new CompletableFuture<>();
+                    promise.completeExceptionally(t);
+                    return (CompletionStage)promise;
+                }
+            }).toCompletableFuture();
+            completeFailedFuture(new TimeoutException(), fallbackMethod, future);
+            return future;
+        } else {
+            return handleProceedWithException(invocation, timeLimiter, fallbackMethod);
```

Comment:
In our Spring Boot TimeLimiterAspect we throw an exception if the return type is not supported.
We currently do not support a usage of the `@TimeLimiter` annotation and a basic return types.
It could be somehow confusing to a developer that the method is "implicitly" executed in a thread pool, but the thread is blocking. I think it would be better if the developer decides it explicitly and uses a ` @Bulkhead(type = Type.THREADPOOL)` or Promise.

Reply:
@RobWin that make sense - I can fix that up